File: C++/binary-search-tree-iterator.cpp
/**
 *  * Definition for binary tree
 *   * struct TreeNode {
 *    *     int val;
 *     *     TreeNode *left;
 *      *     TreeNode *right;
 *       *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 *        * };
 *         */
class BSTIterator {
private:
    std::stack<TreeNode*> stk;
    void go_left(TreeNode* node) {
        while(node) {
            stk.push(node);
            node = node->left;
        }
    }
public:
    BSTIterator(TreeNode *root) {
        go_left(root);
    }

    /** @return whether we have a next smallest number */
    bool hasNext() {
        return !stk.empty();
    }

    /** @return the next smallest number */
    int next() {
        
        auto top = stk.top();
        stk.pop();
        go_left(top->right);
        return top->val;
    }
};

/**
 *  * Your BSTIterator will be called like this:
 *   * BSTIterator i = BSTIterator(root);
 *    * while (i.hasNext()) cout << i.next();
 *     */

File: Python/binary-search-tree-iterator.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class BSTIterator:
    # @param root, a binary search tree's root node
    def __init__(self, root):
        self.stk= []
        self.go_left(root)
        
    def go_left(self, node):
        while node:
            self.stk.append(node)
            node = node.left

    # @return a boolean, whether we have a next smallest number
    def hasNext(self):
        return self.stk

    # @return an integer, the next smallest number
    def next(self):
        top = self.stk.pop()
        self.go_left(top.right)
        return top.val

# Your BSTIterator will be called like this:
# i, v = BSTIterator(root), []
# while i.hasNext(): v.append(i.next())
